hazelcast/src/main/java/com/hazelcast/map/impl/querycache/subscriber/SubscriberAccumulator.java

```diff
@@ -133,17 +119,29 @@ private void removeFromBrokenSequences(QueryCacheEventData event) {
     }
 
     private void handleUnexpectedEvent(QueryCacheEventData event) {
-        addEventSequenceToBrokenSequences(event);
-        publishEventLost(context, info.getMapName(), info.getCacheId(), event.getPartitionId());
-    }
+        // first add sequence of this unexpected event to broken-sequences
+        int partitionId = event.getPartitionId();
+        long sequence = event.getSequence();
+        Long prev = brokenSequences.putIfAbsent(partitionId, sequence);
+
+        // when prev != null, this means we already notified before
+        if (prev != null) {
+            return;
+        }
 
-    private void addEventSequenceToBrokenSequences(QueryCacheEventData event) {
-        Long prev = brokenSequences.putIfAbsent(event.getPartitionId(), event.getSequence());
+        // notify users' EventLostListener
+        InternalQueryCache queryCache = getQueryCache();
+        if (queryCache != null) {
+            if (logger.isWarningEnabled()) {
+                long currentSequence = sequenceProvider.getSequence(partitionId);
+                logger.warning(format("Event lost detected for queryCache=`%s`: "
+                                + "partitionId=%d, expectedSequence=%d, "
+                                + "foundSequence=%d, cacheSize=%d",
+                        queryCache.getCacheId(), partitionId,
+                        currentSequence + 1L, sequence, queryCache.size()));
+            }
 
-        if (prev == null && logger.isFinestEnabled()) {
-            logger.finest(format("Added unexpected event sequence to broken sequences "
-                            + "[partitionId=%d, expected-sequence=%d, broken-sequences-size=%d]",
-                    event.getPartitionId(), event.getSequence(), brokenSequences.size()));
+            publishEventLost(context, info.getMapName(), info.getCacheId(), event.getPartitionId());
```

Comment:
Hmm, I think this is still a change in behavior, isn't it? Before we always called `publishEventLost()` in `handleUnexpectedEvent()` (independently of the `addEventSequenceToBrokenSequences()` method).

Now we can return from `handleUnexpectedEvent()` early, which prevents the logging. But also skips the `publishEventLost()`.

If this is wanted, I'm fine with it. But if users need an event per lost event sequence, we should move this call before the `return`.

Reply:
Yes that is wanted. In the end, we will have one log and one listener notification. This is ok.